memcnn/models/affine.py

```diff
@@ -0,0 +1,563 @@
+        *weights : TorchTensor
+            weights for Fm and Gm in that order {Fm_w1, ... Fm_wn, Gm_w1, ... Gm_wn}
+
+        Note
+        ----
+        All tensor/autograd variable input arguments and the output are
+        TorchTensors for the scope of this fuction
+
+        """
+        # check if possible to partition into two equally sized partitions
+        assert (y.shape[1] % 2 == 0)  # assert if possible
+
+        # store partition size, Fm and Gm functions in context
+        cty.Fm = Fm
+        cty.Gm = Gm
+
+        with torch.no_grad():
+            # partition in two equally sized set of channels
+            y1, y2 = torch.chunk(y, 2, dim=1)
+            y1, y2 = y1.contiguous(), y2.contiguous()
+
+            # compute outputs
+            y1var = y1
+            gmr1, gmr2 = Gm.forward(y1var)
+
+            x2 = (y2 - gmr2) / gmr1
+            y2.set_()
+            del y2
+            x2var = x2
+            fmr1, fmr2 = Fm.forward(x2var)
+            x1 = (y1 - fmr2) / fmr1
+            y1.set_()
+            del y1
+            output = torch.cat([x1, x2], dim=1)
+            x1.set_()
+            del x1
+            x2.set_()
+            del x2
+
+        # save the input and output variables
+        cty.save_for_backward(y.data, output)
+
+        return output
+
+    @staticmethod
+    def backward(cty, grad_output):
+        Fm, Gm = cty.Fm, cty.Gm
+        # are all variable objects now
+        y, output = cty.saved_tensors
+
+        with set_grad_enabled(False):
+            x1, x2 = torch.chunk(output, 2, dim=1)
+            x1, x2 = x1.contiguous(), x2.contiguous()
+
+            # partition output gradient also on channels
+            assert (grad_output.shape[1] % 2 == 0)
+            x1_grad, x2_grad = torch.chunk(grad_output, 2, dim=1)
+            x1_grad, x2_grad = x1_grad.contiguous(), x2_grad.contiguous()
+
+        # Recreate computation graphs for functions Gm and Fm with gradient collecting leaf nodes:
+        # z1_stop, y1_stop, GW, FW
+        # Also recompute inputs (y1, y2) from outputs (x1, x2)
+        with set_grad_enabled(True):
+            z1_stop = x2
+            z1_stop.requires_grad = True
+
+            F_z11, F_z12 = Fm.forward(z1_stop)
+            y1 = x1 * F_z11 + F_z12
+            y1_stop = y1.detach()
+            y1_stop.requires_grad = True
+
+            G_y11, G_y12 = Gm.forward(y1_stop)
+            y2 = x2 * G_y11 + G_y12
+            y2_stop = y2.detach()
+            y2_stop.requires_grad = True
+
+            # restore input
+            y.set_(torch.cat([y1, y2], dim=1).contiguous())
+
+            # compute outputs building a sub-graph
+            z1 = (y2_stop - G_y12) / G_y11
+            x1_ = (y1_stop - F_z12) / F_z11
+            x2_ = z1
+
+            # calculate the final gradients for the weights and inputs
+            dd = torch.autograd.grad(x1_, (z1_stop,) + tuple(Fm.parameters()), x1_grad)
+            z1_grad = dd[0] + x2_grad  # + or - ?
```

Comment:
please try this against the new unittests also for the other Affine implementations
```suggestion
            z1_grad = dd[0] - x2_grad 
```